File: python/590.py

# Definition for a Node.
from collections import deque
from typing import List


class Node:
  def __init__(self, val=None, children=None):
    self.val = val
    self.children = children

# 递归
class Solution:
  def postorder(self, root: 'Node') -> List[int]:
    ans = []
    def helper(node):
      if not node: return
      for child in node.children:
        helper(child)
      ans.append(node.val)
    helper(root)
    return ans

# 迭代
class Solution:
  def postorder(self, root: 'Node') -> List[int]:
    ans = []
    q = deque([root])
    while q and q[-1]:
      node = q.popleft()
      ans.append(node.val)
      # for i in range(len(node.children)):
      #   q.appendleft(i)
      q.extendleft(node.children)
    return ans[::-1]

File: go/590.go
package main

// func postorder(root *Node) (ans []int) {
// 	var helper func(node *Node)
// 	helper = func(node *Node) {
// 		for _, child := range node.Children {
// 			helper(child)
// 		}
// 		ans = append(ans, node.Val)
// 	}
// 	if root == nil {
// 		return
// 	}
// 	helper(root)
// 	return
// }

func postorder(root *Node) (ans []int) {
	if root == nil {
		return
	}
	st := []*Node{root}
	for len(st) > 0 {
		ans = append(ans, st[len(st)-1].Val)
		childs := st[len(st)-1].Children
		st = st[:len(st)-1]
		st = append(st, childs...)
	}
	for i, j := 0, len(ans)-1; i < j; i, j = i+1, j-1 {
		ans[i], ans[j] = ans[j], ans[i]
	}
	return
}
